src/ophys_etl/decrosstalk/decrosstalk_types.py

```diff
@@ -0,0 +1,334 @@
+from typing import List, Tuple
+import numpy as np
+
+
+class BasicDictWrapper(object):
+
+    def __contains__(self, key) -> bool:
+        if key in self._data:
+            return True
+        return False
+
+    def __len__(self) -> int:
+        return len(self._data)
+
+    def __iter__(self):
+        raise NotImplementedError("Did not implement iterator over "
+                                  "BasicDictWrapper")
+
+################## classes for handling pure traces #############  # noqa: E266
+
+
+class ROIChannels(BasicDictWrapper):
+    """
+    A wrapper around a dict meant to store all of the decrosstalk-related
+    data for one ROI.
+
+    Valid key, value pairs are:
+
+    'signal' -- a 1-D np.ndarray of floats containing a trace
+    'crosstalk' -- a 1-D np.ndarray of floats containing a trace
+    'mixing_matrix' -- a 2x2 np.ndarray of floats
+    'poorly_converged_signal' -- same type as 'signal'
+    'poorly_converged_crosstalk' -- same type as 'crosstalk'
+    'poorly_converged_mixing_matrix' -- same type as 'mixing_matrix'
+    'use_avg_mixing_matrix' -- a boolean
+    """
+
+    def __init__(self):
+        self._data = {}
+        self._valid_keys = ('signal',
+                            'crosstalk',
+                            'mixing_matrix',
+                            'poorly_converged_signal',
+                            'poorly_converged_crosstalk',
+                            'poorly_converged_mixing_matrix',
+                            'use_avg_mixing_matrix')
+
+    def _validate_np_array(self,
+                           key: str,
+                           value: np.ndarray,
+                           expected_ndim: int = None,
+                           expected_shape: Tuple = None,
+                           expected_dtype: np.dtype = None) -> bool:
+
+        if not isinstance(value, np.ndarray):
+            msg = '%s must be a np.ndarray; you gave %s\n' % (key, type(value))
```

Comment:
I recommend python3 string interpolation

Reply:
I'm honestly asking: what is your reason for preferring that? It always makes me a little nervous trusting python to grab the right thing out of the current context.

Reply:
I think of it more that you are trusting the writers of the object to specify the `__format__` and then there are the pythonic/aesthetic arguments:
https://realpython.com/python-f-strings/

Reply:
or rather, I guess it is the `__str__` and `__repr__` methods:
```
In [14]: x = {'a': 1}                                                                                         

In [15]: x.__str__()                                                                                          
Out[15]: "{'a': 1}"

In [16]: x.__repr__()                                                                                         
Out[16]: "{'a': 1}"

Reply:
Okay. I'll keep that in mind going forward. I'm going to go ahead and merge this now for the sake of speed. Thanks for the insight.